### PR TITLE
sql: Add new builtin generator function `crdb_internal.scan_storage_internal_keys()`

### PR DESCRIPTION
**Pull request by raggar**
This new builtin is used to gather specific pebble metrics for a node and store id (within an given keyspan). The builtin returns information about the different types of keys (including snapshot pinned keys) and bytes.

Informs: #94659
Release-note: None

### CODE REVIEW

File: pkg/storage/pebble.go [hunk 1]
```diff
@@ -2063,6 +2063,15 @@ func (p *Pebble) GetTableMetrics(start, end roachpb.Key) ([]enginepb.SSTableMetr
 	return metricsInfo, nil
 }
 
+// GetAdvancedPebbleMetrics implements the Engine interface.
+func (p *Pebble) GetAdvancedPebbleMetrics(
+	start, end roachpb.Key,
+) ([]enginepb.AdvancedPebbleMetrics, error) {
+	fmt.Println("START", start)
+	fmt.Println("END", end)
+	return []enginepb.AdvancedPebbleMetrics{}, nil
```
Comment (RahulAggarwal1016) [inline pkg/storage/pebble.go:2072]: once https://github.com/cockroachdb/pebble/pull/2713 is merged, I will call `ScanStatistics` to retrieve the metrics.

File: pkg/storage/pebble.go [hunk 2]
```diff
@@ -2063,6 +2063,13 @@ func (p *Pebble) GetTableMetrics(start, end roachpb.Key) ([]enginepb.SSTableMetr
 	return metricsInfo, nil
 }
 
+// GetAdvancedPebbleMetrics implements the Engine interface.
+func (p *Pebble) GetAdvancedPebbleMetrics(
+	start, end roachpb.Key,
+) ([]enginepb.AdvancedPebbleMetrics, error) {
+	return []enginepb.AdvancedPebbleMetrics{}, nil
```
Comment (RahulAggarwal1016) [inline pkg/storage/pebble.go:2070]: once https://github.com/cockroachdb/pebble/pull/2713 is merged, I will call ScanStatistics to retrieve the metrics.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:  Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-:-NagT0Ye-oSPtKJrsLd7:b-bsmgd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @RahulAggarwal1016, and @rharding6373)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagTCrK9OZKvzggDmhr:-NagTCrK9OZKvzggDmhs:b-c8npgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L687)):*
> ```Go
> 		),
> 	),
> 	"crdb_internal.pebble_metrics": makeBuiltin(
> ```

@jbowens any opinions about the name? `pebble_key_metrics` or `sstable_key_metrics` are possible alternatives
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 701 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagSgOF9vsXwaB8d9h6:-NagSgOF9vsXwaB8d9h7:bexd57k) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L701)):*
> ```Go
> 			makeAdvancedPebbleMetricsGenerator,
> 			"Returns statistics for a pebble store in the range start_key and end_key for the provided node id.",
> 			volatility.Stable,
> ```

This (and sstable_metrics) should be `volatility.Volatile`. Stable means it returns the same result when executed multiple times in the same txn (not a big deal though given its typical usage).

**Review by jbowens — COMMENTED** [review]
> Reviewed 1 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @RaduBerinde, @RahulAggarwal1016, and @rharding6373)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagjKVFEHBMxiRsIzKv):*
nit: prefix with `sql,kv:` and use imperative voice in the commit title: "Add new builtin ..."
___
*[`pkg/kv/kvserver/stores_server.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagkVNf4ZM82x19e5uk:-NagkVNf4ZM82x19e5ul:babj2ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/kv/kvserver/stores_server.go#L176)):*
> ```Go
> }
> 
> func (is Server) GetAdvancedPebbleMetrics(
> ```

Throughout let's remain consistent with the SQL function name we decide, eg ScanStorageInternalKeys
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagTCrK9OZKvzggDmhr:-NagjgKq0xa9L7mSFGkQ:b-w5p2z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L687)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

@jbowens any opinions about the name? `pebble_key_metrics` or `sstable_key_metrics` are possible alternatives
</blockquote></details>

How about `scan_storage_internal_keys` — I think including "scan" is appropriate to convey the cost of the operation.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagjrhJ4d7vwgQPTQqj:-NagjrhJ4d7vwgQPTQqk:b-2to5o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L700)):*
> ```Go
> 			advancedPebbleMetricsGeneratorType,
> 			makeAdvancedPebbleMetricsGenerator,
> 			"Returns statistics for a pebble store in the range start_key and end_key for the provided node id.",
> ```

How about: "Scans a store's storage engine, computing statistics describing the internal keys within the span [start_key, end_key)."
___
*[`pkg/storage/pebble.go` line 2070 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#gh-1279343220:-NagkhCX1nr6jteezcvf:b6maqm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/storage/pebble.go#L2070)):*
<details><summary><i>Previously, RahulAggarwal1016 (Rahul Aggarwal) wrote…</i></summary><blockquote>

once https://github.com/cockroachdb/pebble/pull/2713 is merged, I will call ScanStatistics to retrieve the metrics.
</blockquote></details>

Let's hold off merging this PR until that's in place in case the statistics we compute change.

**Review by raggar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, @RaduBerinde, and @rharding6373)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagTCrK9OZKvzggDmhr:-NagtsjJEfnZ4Gl_5CAi:b-fcb4ag) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L687)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

How about `scan_storage_internal_keys` — I think including "scan" is appropriate to convey the cost of the operation.
</blockquote></details>

yeah that sounds good

**Review by raggar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @jbowens, @RaduBerinde, and @rharding6373)

___
*[`pkg/storage/pebble.go` line 2070 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#gh-1279343220:-NagtwhD1axePUmeWOd3:b-tl2y4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/storage/pebble.go#L2070)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Let's hold off merging this PR until that's in place in case the statistics we compute change.
</blockquote></details>

yup thats the plan

**Review by raggar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bananabrick, @jbowens, @RaduBerinde, and @rharding6373)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagjKVFEHBMxiRsIzKv):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: prefix with `sql,kv:` and use imperative voice in the commit title: "Add new builtin ..."
</blockquote></details>

done
___
*[`pkg/kv/kvserver/stores_server.go` line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagkVNf4ZM82x19e5uk:-Nah6pLy1ARxcUTVwJxm:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/kv/kvserver/stores_server.go#L176)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Throughout let's remain consistent with the SQL function name we decide, eg ScanStorageInternalKeys
</blockquote></details>

fixed
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagjrhJ4d7vwgQPTQqj:-Nah6rzf4sdAlTKUyo5k:bypp4it) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L700)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

How about: "Scans a store's storage engine, computing statistics describing the internal keys within the span [start_key, end_key)."
</blockquote></details>

that sounds good
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 701 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NagSgOF9vsXwaB8d9h6:-Nah6r1uBLu76SAoJwl8:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd6c32423fd754ef692ace76c8f23c3fcb2d5376/pkg/sql/sem/builtins/generator_builtins.go#L701)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This (and sstable_metrics) should be `volatility.Volatile`. Stable means it returns the same result when executed multiple times in the same txn (not a big deal though given its typical usage).
</blockquote></details>

fixed

**Review by rharding6373 — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-:-Nalsw_WD569q7JyDOM1:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @bananabrick, @jbowens, and @RaduBerinde)

**Comment by RaduBerinde** [comment]
*[`pkg/sql/sem/builtins/generator_builtins.go` line 697 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NawDYJD2DI50izFnxvF:-NawDYJD2DI50izFnxvG:bgiuwki) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b19e7b65bb56fb5768e7e8f1f8f161ff05148f9/pkg/sql/sem/builtins/generator_builtins.go#L697)):*
> ```Go
> 				{Name: "start_key", Typ: types.Bytes},
> 				{Name: "end_key", Typ: types.Bytes},
> 			},
> ```

I know there's a plan to add an argument for rate limiting. We should have two overloads, one with and one without the argument (essentially making that argument optional). The argument should be in megabytes / second for convenience. The one without the argument should use a value on the low side, say 10MB/s - it should be "safe" to use in all cases, and the user can increase the rate if it's too slow.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @bananabrick, @jbowens, and @RahulAggarwal1016)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 713 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NbG4u8mBASiYQZBq2S1:-NbG4u8mBASiYQZBq2S2:b-bryccz) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e31f137840a5a4526ce36c8e2bd41e040d1d07/pkg/sql/sem/builtins/generator_builtins.go#L713)):*
> ```Go
> 			storageInternalKeysGeneratorType,
> 			makeStorageInternalKeysGenerator,
> 			"Scans a store's storage engine, computing statistics describing the internal keys within the span [start_key, end_key). This funct",
> ```

[nit] unfinished description

File: pkg/storage/pebble.go [hunk 3]
```diff
@@ -2072,6 +2072,44 @@ func (p *Pebble) GetTableMetrics(start, end roachpb.Key) ([]enginepb.SSTableMetr
...
+			PointKeyDeleteCount:     uint64(stats.Levels[level].KindsCount[pebble.InternalKeyKindDelete]),
+			PointKeySetCount:        uint64(stats.Levels[level].KindsCount[pebble.InternalKeyKindSet]),
+			RangeDeleteCount:        uint64(stats.Levels[level].KindsCount[pebble.InternalKeyKindRangeDelete]),
+			RangeKeySetCount:        uint64(stats.Levels[level].KindsCount[pebble.InternalKeyKindRangeKeySet]),
+			RangeKeyDeleteCount:     uint64(stats.Levels[level].KindsCount[pebble.InternalKeyKindRangeKeyDelete]),
+		})
+	}
+
+	metrics = append(metrics, enginepb.StorageInternalKeysMetrics{
+		Level:                   -1,
```
Comment (RahulAggarwal1016) [inline pkg/storage/pebble.go:2100]: not sure what value this should be (is there a way to have a column be blank?)

**Comment by RaduBerinde** [comment]
*[`pkg/storage/pebble.go` line 2100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107743#gh-1290234203:-NbUtYZM8sjuEwSfOiTE:ba8oc60) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d132283810e3ea52b5ba7f745f1bda5849cfe7/pkg/storage/pebble.go#L2100)):*
<details><summary><i>Previously, RahulAggarwal1016 (Rahul Aggarwal) wrote…</i></summary><blockquote>

not sure what value this should be (is there a way to have a column be blank?)
</blockquote></details>

You can convert -1 to DNull when you generate the datums

**Comment by RahulAggarwal1016** [comment]
> _[`pkg/storage/pebble.go` line 2100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/107743#gh-1290234203:-NbUtYZM8sjuEwSfOiTE:ba8oc60) ([raw file](https://github.com/cockroachdb/cockroach/blob/21d132283810e3ea52b5ba7f745f1bda5849cfe7/pkg/storage/pebble.go#L2100)):_
> 
> Previously, RahulAggarwal1016 (Rahul Aggarwal) wrote…
> You can convert -1 to DNull when you generate the datums

oh nice yeah i'll do that

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-:-NbUzdxtCRJV0O3KFfPE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @bananabrick, @jbowens, and @RahulAggarwal1016)

___
*[`pkg/kv/kvserver/api.proto` line 92 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/107743#-NbUzR6K3kpvfXjbetRR:-NbUzR6K3kpvfXjbetRS:b9b6ts0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2bf9003ba7c446bdbb350653b7fec8b2e335f01/pkg/kv/kvserver/api.proto#L92)):*
> ```proto
> }
> 
> message ScanStorageInternalKeysRequest {
> ```

[nit] would be good to have a brief comment here

**Pull request merged**